### ISSUE TITLE
[SR][UX] Provide more guidance on password strength required

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#5095]
**Description of the need**

Currently the description under the password field states that "Weak passwords will be rejected".

![current-password-description](https://user-images.githubusercontent.com/1071483/118522952-a45c4b00-b70a-11eb-8890-d547b64c0ab6.png)

I've gotten feedback that it's not clear how strong the password needs to be in order to be accepted. What is "weak" in this context? Plus the actual threshold can be changed in config: `user_password_strength_threshold` defaults to "50" ("fair") but can be increased or decreased so that "weak", "fair", "good" or "excellent" passwords are required.

**Proposed solution**

Make the description dynamic so that it reflects the actual threshold set in config. For example:

![proposed-password-description](https://user-images.githubusercontent.com/1071483/118522969-a920ff00-b70a-11eb-817b-391d5641446f.png)

**Additional information**

This is a follow-up to https://github.com/backdrop/backdrop-issues/issues/4265

**Comment by indigoxela** [comment]
> I've gotten feedback that it's not clear how strong the password needs to be in order to be accepted.

Fair enough, it's really a bit vague. Even more if the threshold has been adapted. I'm confident, that our "UX task force" will take care of the details.

Off topic: are there any plans yet for an admin interface for user_password_strength_threshold? Or is that feature hidden by concept? If we already have an issue for that - I either forgot or totally missed it.

**Comment by herbdool** [comment]
@indigoxela I do have a PR if you're interested in testing.

I didn't have plans for an admin interface for user_password_strength_threshold. It could be left to contrib. There's also a hook so that might mean contrib could make it even more dynamic: different for each role, for example.

**Comment by indigoxela** [comment]
> I do have a PR if you're interested in testing.

It's not urgent, I'm short on time anyway.

> There's also a hook so that might mean contrib could make it even more dynamic: different for each role, for example.

Do you mean hook_user_password_reject_weak_alter()? I think that only alters validation - after form submission. Or does that also have influence on the dynamic help text?

**Comment by klonos** [comment]
> ...If we already have an issue for that - I either forgot or totally missed it.

That setting could live under #3624

**Comment by klonos** [comment]
@herbdool I've left a couple of comments in the PR, but generally looks good. Thanks 👍🏼

**Comment by herbdool** [comment]
@indigoxela yes, that's the one (I was just going by memory before). I think `user_password_reject_weak()` and the alter hook just decide if the password should be checked or not for a particular user. And the help text only shows if it's true. So maybe that's good enough.

@klonos thanks. I kinda like that we're not exposing that setting in the UI, because most people won't need/want to change it.

**Comment by indigoxela** [comment]
I'll try to do some local testing with a custom hook implementation soon - out of curiosity and for better understanding.

I think the dynamic text is helpful for users. :+1: 

Re the admin interface for setting user_password_strength_threshold:
That could be part of another core issue, but I also get the point of @herbdool - "most people won't need/want to change it" - I agree, this might as well be a task for contrib.

**Comment by indigoxela** [comment]
Works for me (except for some silly interface text, which isn't a blocker for me). Code review should be done by someone else. :wink:

**Comment by klonos** [comment]
Code LGTM in general, but the logic could be improved. For instance, I see this bit being repeated twice:

```php
    if ($reject_weak) {
      $description .= ' ' . t('Password must be <em>@threshold</em> or stronger.', array('@threshold' => $threshold_label));
      $description .= _user_password_policy_help();
    }
```

Any reason why we couldn't (conditionally) include the "Password must be..." text in `_user_password_policy_help()`, so that it's part of the `<ul>` that that function returns? If this entire conditional + also the `$threshold_label` logic get moved to `_user_password_policy_help()`, we wouldn't be calling `t()` twice on the same string, and that code duplication would be simplified to a single line: `$description .= _user_password_policy_help();`

Also, for consistency, the string should start with "The password must be...". And lastly, the threshold help text should always be shown in the "Account settings" form (for admins) - it doesn't as it is now.

**Comment by klonos** [comment]
@herbdool here's a PR against yours: https://github.com/herbdool/backdrop/pull/14

It changes this:
![image](https://user-images.githubusercontent.com/2423362/119272121-2d90e780-bc0d-11eb-9ed6-856f7b7e26a0.png)

...to this:
![image](https://user-images.githubusercontent.com/2423362/119273526-04278a00-bc14-11eb-8d66-f92e04c68dc4.png)

And also this:
![image](https://user-images.githubusercontent.com/2423362/119273149-30daa200-bc12-11eb-8381-5ac5ccbdd4e2.png)

...to this instead:
![image](https://user-images.githubusercontent.com/2423362/119273506-ece89c80-bc13-11eb-9aac-af03175080da.png)

**Comment by herbdool** [comment]
I've incorporated @klonos suggested changes with some tweaks. I had to fix the CSS class since it was going to be a translated word which could easily break the CSS.

Also, I do not think we should confuse the indicator threshold for the criteria. The calculation takes into account: username, email, length and complexity. The *output* of that is a number and it must pass a threshold. That threshold cannot be the criteria itself.

**Comment by klonos** [comment]
Thanks @herbdool 🙏🏼 ...good call on the CSS class 👍🏼 

Tested quickly and the only thing that I'd like us to fix is to hide the "The current password must be entered to set a new password." from the "New password" field when a user is being edited by admins:

<img width="70%" src="https://user-images.githubusercontent.com/2423362/120472054-20fb5480-c3ae-11eb-94e1-a41466119e68.png">

Since admins are allowed to update other users' passwords without entering their current password, this help text doesn't make sense in that case (there's no "Current password" field rendered in that form anyway).

Of course, that help text should remain in place when a user is editing their own password:

<img width="70%" src="https://user-images.githubusercontent.com/2423362/120472545-af6fd600-c3ae-11eb-9e3a-502d759e1fde.png">

**Comment by herbdool** [comment]
Both @klonos and I worked on the code, so would be good to get someone else to review the code. @indigoxela are you up for it?

**Comment by indigoxela** [comment]
> Both @klonos and I worked on the code

@herbdool I see, yes, in this case I should do the review.

Code looks good, RTBC from my side. Many thanks for working on this. :+1:

**Comment by quicksketch** [comment]
I have merged https://github.com/backdrop/backdrop/pull/3628 into 1.x, thanks so much @herbdool, @indigoxela, and @klonos!

Should this be pulled into 1.19.2? As a UX improvement seems like it could be included any time.

**Comment by indigoxela** [comment]
> Should this be pulled into 1.19.2? As a UX improvement seems like it could be included any time.

IMO it could easily go into 1.19.2. @herbdool you probably set the label back from task to "feature request" for a good reason - mind to explain?

**Comment by herbdool** [comment]
It can be 1.19.2. I don't often think about difference between feature and task, I chose one that seemed close.

**Comment by klonos** [comment]
Thanks @herbdool ...I think that we should merge it into 1.19.x as well then.

**Comment by jenlampton** [comment]
> Should this be pulled into 1.19.2? 

Since 1.19.3 is coming out today with this fix, I'm going to mark this issue as fixed and close it.

**Comment by herbdool** [comment]
@jenlampton was this actually merged into 1.19.x as well? I only see it in 1.x.

**Comment by quicksketch** [comment]
Sorry folks. I seem to have only merged this into 1.x. I cherry-picked the commit back into 1.19.x as well today.
